zcash_client_backend/src/proto.rs

```diff
@@ -177,3 +196,249 @@ impl service::TreeState {
         read_commitment_tree::<Node, _, NOTE_COMMITMENT_TREE_DEPTH>(&sapling_tree_bytes[..])
     }
 }
+
+pub const PROPOSAL_SER_V1: u32 = 1;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ProposalError<DbError> {
+    Zip321(Zip321Error),
+    TxIdInvalid(Vec<u8>),
+    InputRetrieval(DbError),
+    BalanceInvalid,
+    MemoInvalid(memo::Error),
+    VersionInvalid(u32),
+    ZeroMinConf,
+    FeeRuleNotSpecified,
+}
+
+impl<E> From<Zip321Error> for ProposalError<E> {
+    fn from(value: Zip321Error) -> Self {
+        Self::Zip321(value)
+    }
+}
+
+impl proposal::ProposedInput {
+    pub fn parse_txid(&self) -> Result<TxId, Vec<u8>> {
+        Ok(TxId::from_bytes(self.txid.clone().try_into()?))
+    }
+}
+
+impl proposal::Proposal {
+    /// Serializes a [`Proposal`] based upon a supported [`StandardFeeRule`] to its protobuf
+    /// representation.
+    pub fn from_standard_proposal<P: Parameters, NoteRef>(
+        params: &P,
+        value: &Proposal<StandardFeeRule, NoteRef>,
+    ) -> Option<Self> {
+        let transaction_request = value.transaction_request().to_uri(params)?;
+        let inputs = value
+            .sapling_inputs()
+            .iter()
+            .map(|rec_note| proposal::ProposedInput {
+                txid: rec_note.txid().as_ref().to_vec(),
+                pool: 2,
```

Comment:
We should have constants for these.

Reply:
Is this a pool code (`zcash_client_sqlite::wallet::pool_code`) or a typecode (`u32::from(zcash_address::kind::unified::Typecode))`)?

Reply:
Is this a pool code (`zcash_client_sqlite::wallet::pool_code`) or a typecode (`u32::from(zcash_address::kind::unified::Typecode))`)? It doesn't appear to be a `proposal::ValuePool`.

Too many similar concepts. Let's just use ZIP 316 typecodes everywhere.

Reply:
This has been removed in favor of storing separate input data structures, which is necessary if Sapling and Orchard parts of the proposal use different anchors anyway.